swim/memberlist.go

```diff
@@ -275,16 +278,138 @@ func (m *memberlist) MakeFaulty(address string, incarnation int64) []Change {
 }
 
 func (m *memberlist) SetLocalStatus(status string) {
+	m.members.Lock()
 	m.local.Status = status
+	m.members.Unlock()
 	m.postLocalUpdate()
 }
 
+func (m *memberlist) SetLocalLabel(key, value string) error {
+	// TODO implement a sane limit for the size of the labels to prevent users
+	// from impacting the performance of the gossip protocol.
+
+	m.members.Lock()
+
+	// ensure that there is a labels map
+	if m.local.Labels == nil {
+		m.local.Labels = make(map[string]string)
+	}
+
+	old, had := m.local.Labels[key]
+
+	// set the label
+	m.local.Labels[key] = value
+	m.members.Unlock()
+
+	if !had || old != value {
+		// postLocalUpdate reincarnates and starts gossipping the new state
+		// which is only desired when a change to the local labels has been made
+		m.postLocalUpdate()
+	}
+
+	// there was no error during the manipulation of labels
+	return nil
+}
+
+// GetLocalLabel returns the value of a label set on the local node. Its second
+// argument indicates if the key was present on the node or not
+func (m *memberlist) GetLocalLabel(key string) (string, bool) {
+	m.members.RLock()
+	value, has := m.local.Labels[key]
+	m.members.RUnlock()
+	return value, has
+}
+
+// LocalLabelsAsMap copies the labels set on the local node into a map for the
+// callee to use. Changes to this map will not be reflected in the labels kept
+// by this node.
+func (m *memberlist) LocalLabelsAsMap() map[string]string {
+	m.members.RLock()
+	defer m.members.RUnlock()
+	if len(m.local.Labels) == 0 {
+		return nil
+	}
+
+	cpy := make(map[string]string, len(m.local.Labels))
+	for k, v := range m.local.Labels {
+		cpy[k] = v
+	}
+	return cpy
+}
+
+// SetLocalLabels updates multiple labels at once. It will take all the labels
+// that are set in the map passed to this function and overwrite the value with
+// the value in the map. Keys that are not present in the provided map will
+// remain in the labels of this node. The operation is guaranteed to succeed
+// completely or not at all.
+func (m *memberlist) SetLocalLabels(labels map[string]string) error {
+	m.members.Lock()
+	// ensure that there is a labels map
+	if m.local.Labels == nil {
+		m.local.Labels = make(map[string]string, len(labels))
+	}
+
+	// keep track if we made changes to the labels
+	changes := false
+
+	// copy the key-value pairs to our internal labels. By not setting the map
+	// of labels to the Labels value of the local member we prevent removing labels
+	// that the user did not specify in the new map.
+	for key, value := range labels {
+		old, had := m.local.Labels[key]
+		m.local.Labels[key] = value
+
+		if !had || old != value {
+			changes = true
+		}
+	}
+
+	m.members.Unlock()
+
+	if changes {
+		m.postLocalUpdate()
+	}
+
+	return nil
+}
+
+// Remove a label from the local map of labels. This will trigger a reincarnation
```

Comment:
s/a label/a list of labels/

Also, mention that it is not an error to call this function with a non-existing key.